maxi-blocks/maxi-blocks
# Issue 3489 - `uniqueID` order generator

# Description

Basically, the older method before the implementation of #2765 was creating this long `uniqueID` numbers that then has been converted to the new custom-labels. To even improve it a bit more, this PR proposes to stop depending on an uncontrolled function to set the number and starts basing on the current actual number of blocks of the same type on the post. This way, we aren't going to have long numbers 👍 The solution implemented is based in 2 parts:
1. `uniqueIDGenerator`: that is going to ensure that numeration.
2. This snippet that is going to organize the current created patterns:
```jsx
(() => {
	const { getBlocks } = wp.data.select('core/block-editor');
	const { updateBlockAttributes } = wp.data.dispatch('core/block-editor');

	const blocks = getBlocks();

	const blocksType = {};

	const uniqueIDGenerator = (blockName, diff = 0) => {
		const newID = `${blockName
			.replace('maxi-blocks/', '')
			.replace(/[0-9]/g, '')}-${blocksType[blockName] + diff}`;

		return newID;
	};

	const blockIterator = blocks => {
		blocks.forEach(block => {
			if (block.innerBlocks.length) {
				blockIterator(block.innerBlocks);
			}
			setTimeout(() => {
				if (blocksType[block.name]) blocksType[block.name] += 1;
				else blocksType[block.name] = 1;

				const currentUniqueID = block.attributes.uniqueID;

				const uniqueID = uniqueIDGenerator(block.name);

				const needNewLabel = (() => {
					const label = currentUniqueID.replace('-maxi-', '_');

					const customLabel =
						label.charAt(0).toUpperCase() + label.slice(1);

					return customLabel === block.attributes.customLabel;
				})();

				const label = uniqueID.replace('-maxi-', '_');
				const customLabel =
					label.charAt(0).toUpperCase() + label.slice(1);

				updateBlockAttributes(block.clientId, {
					uniqueID,
					...(needNewLabel && { customLabel }),
				});
			}, 200);
		});
	};

	blockIterator(blocks);
})();
```

Fixes #3489 

# How Has This Been Tested?

The testing comes in 2 different parts:
1. Create a post, add blocks (no pasting, manually) and check the numeration works correctly. Delete some blocks and add new ones, numeration shouldn't repeat and should return a long number. IMPORTANT: the numeration is not sequential, is based on the number of same type blocks: if we add 7 Group Maxi and remove first 4 ones, the next number for a Group Maxi is going to be 4.
2. Create a post and add some pattern from maxi-dress (here's a [random one](https://github.com/yeahcan/maxi-blocks/files/9139243/random.txt)). Once done, open the console and run the snippet there; the `uniqueID` and the `customLabel` should be organized 👍 

# Test checklist



**_ Front/Back Testing _**

-   [x] Duplicate the block, test that the `uniqueID` and the `customLabel` hasn't long numbers
-   [ ] Test with many blocks of the same type

**_ Pre-Code Testing _**

-   [ ] Duplicate the block, test that the `uniqueID` and the `customLabel` hasn't long numbers
-   [ ] Test with many blocks of the same type

# Checklist
